flocker/apiclient/test/test_client.py

```diff
@@ -0,0 +1,181 @@
+            """
+            If no ``dataset_id`` is specified when calling ``create_dataset``,
+            a new one is generated.
+            """
+            return self.assert_creates(client_factory(), primary=self.node_1,
+                                       maximum_size=DATASET_SIZE)
+
+        def test_create_given_dataset(self):
+            """
+            If no ``dataset_id`` is specified when calling ``create_dataset``,
+            a new one is generated.
+            """
+            return self.assert_creates(client_factory(), primary=self.node_1,
+                                       maximum_size=DATASET_SIZE,
+                                       dataset_id=uuid4())
+
+        def test_create_with_metadata(self):
+            """
+            The metadata passed to ``create_dataset`` is stored with the
+            dataset.
+            """
+            return self.assert_creates(client_factory(), primary=self.node_1,
+                                       maximum_size=DATASET_SIZE,
+                                       metadata={u"hello": u"there"})
+
+        def test_create_conflicting_dataset_id(self):
+            """
+            Creating two datasets with same ``dataset_id`` results in an
+            ``DatasetAlreadyExists``.
+            """
+            client = client_factory()
+            d = self.assert_creates(client, primary=self.node_1,
+                                    maximum_size=DATASET_SIZE)
+
+            def got_result(dataset):
+                d = client.create_dataset(primary=self.node_1,
+                                          maximum_size=DATASET_SIZE,
+                                          dataset_id=dataset.dataset_id)
+                return self.assertFailure(d, DatasetAlreadyExists)
+            d.addCallback(got_result)
+            return d
+
+        def test_move(self):
+            """
+            ``move_dataset`` changes the dataset's primary.
+            """
+            client = client_factory()
+            dataset_id = uuid4()
+
+            d = self.assert_creates(client, primary=self.node_1,
+                                    maximum_size=DATASET_SIZE,
+                                    dataset_id=dataset_id)
+            d.addCallback(
+                lambda _: client.move_dataset(self.node_2, dataset_id))
+
+            def got_result(dataset):
+                listed = client.list_datasets_configuration()
+                listed.addCallback(lambda l: (dataset, l))
+                return listed
+            d.addCallback(got_result)
+
+            d.addCallback(lambda result:
+                          self.assertEqual((result[0], result[0] in result[1]),
+                                           (Dataset(dataset_id=dataset_id,
+                                                    primary=self.node_2,
+                                                    maximum_size=DATASET_SIZE),
+                                            True)))
```

Comment:
I found this assertion difficult to follow. Partly it's because it's combining two assertions into one in order to maximize information in the assertion.

Suggestions to make it easier:
 * unpack `result` in the lambda to `(dataset, found_datasets)` or similar
 * use `@inlineCallbacks`